File: Leetcode/653_Two_Sum_IV_Input_is_a_BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left https://leetcode.com/discuss/= left
#         self.right = right
class Solution(object):
    def findTarget(self, root, k):
        cache = set()
        return self.bst_helper(root, k, cache)
        
    def bst_helper(self, root, target, cache):
        if not root:
            return False
        if target-root.val in cache:
            return True
        cache.add(root.val)
        return self.bst_helper(root.left,target,cache) or self.bst_helper(root.right,target, cache)


        

File: Leetcode/653_Two_Sum_IV_Input_is_a_BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution{
    // tc - N sc -  N set of size can go till N
    public boolean findTarget(TreeNode root, int k) {
        Set<Integer> set = new HashSet<>();
        return helper(root, set,k);        
    }
    
    private boolean helper(TreeNode root, Set<Integer> set, int k){
        if(root==null) return false;
        if(set.contains(k-root.val)) return true; 
        set.add(root.val);
        return helper(root.left, set, k) 
            || helper(root.right, set, k);
    }
}